<div align="center">

# Spatial Augmentations

Graph augmentation strategies for self-supervised pretraining of graph neural networks in spatial omics.

<a href="https://pytorch.org/get-started/locally/"><img alt="PyTorch" src="https://img.shields.io/badge/PyTorch-ee4c2c?logo=pytorch&logoColor=white"></a>
<a href="https://pytorchlightning.ai/"><img alt="Lightning" src="https://img.shields.io/badge/-Lightning-792ee5?logo=pytorchlightning&logoColor=white"></a>
<a href="https://hydra.cc/"><img alt="Config: Hydra" src="https://img.shields.io/badge/Config-Hydra-89b8cd"></a>
<a href="https://github.com/ashleve/lightning-hydra-template"><img alt="Template" src="https://img.shields.io/badge/-Lightning--Hydra--Template-017F2F?style=flat&logo=github&labelColor=gray"></a><br>

</div>

## Description

**Spatial Augmentations** is a research framework for exploring and benchmarking graph-based data augmentation strategies to improve Graph Neural Network (GNN) pretraining on spatial omics datasets (e.g., spatial transcriptomics, spatial proteomics). This project is built on [PyTorch Lightning](https://lightning.ai/) and [Hydra](https://hydra.cc/) for modularity, reproducibility, and ease of experimentation.

The key method used is **Bootstrapped Graph Latents (BGRL)** ([Thakoor et al., 2021](https://arxiv.org/abs/2102.06514)) for self-supervised pretraining of GNNs via graph augmentations. Augmentation strategies are evaluated on two representative tasks: domain identification and phenotype prediction.

This repository is associated with the ETH Zürich semester project:\
**“Exploring Augmentation-Driven Inductive Biases for Pretraining Graph Neural Networks in Spatial Omics”**\
Student: Michel Tarnow\
Supervisors: Lovro Rabuzin, Prof. Dr. Valentina Boeva

## Project Structure

```
├── .github                   <- Github Actions workflows
├── configs                   <- Hydra configs
├── data                      <- Project data
├── logs                      <- Logs generated by hydra and lightning loggers
├── notebooks                 <- Jupyter notebooks
├── scripts                   <- Shell scripts
├── src                       <- Source code
├── tests                     <- Tests of any kind
│
├── .gitignore                <- List of files ignored by git
├── .pre-commit-config.yaml   <- Configuration of pre-commit hooks for code formatting
├── .project-root             <- File for inferring the position of project root directory
├── environment.yaml          <- File for installing conda environment
├── LICENSE                   <- License file
├── Makefile                  <- Makefile with commands like `make train` or `make test`
├── pyproject.toml            <- Configuration options for testing and linting
├── requirements.txt          <- File for installing python dependencies
└── README.md
```

## Data Access

Datasets are not included in this repository due to their size. Raw datasets can be downloaded from the resources provided below.

| Dataset       | Task                  | Description                                                                                                                                                                                         |
| ------------- | --------------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| **Domain123** | Domain Identification | Datasets 1, 2, and 3 from [Schaub et al. (2025)](https://academic.oup.com/bioinformatics/article/41/1/btaf005/7945104): mouse brain spatial transcriptomics datasets (MERFISH, STARmap, BaristaSeq) |
| **Domain4**   | Domain Identification | Datasets 4 from [Schaub et al. (2025)](https://academic.oup.com/bioinformatics/article/41/1/btaf005/7945104): mouse brain spatial transcriptomics dataset (Xenium)                                  |
| **Domain7**   | Domain Identification | Dataset 7 from [Schaub et al. (2025)](https://academic.oup.com/bioinformatics/article/41/1/btaf005/7945104): mouse brain spatial transcriptomics dataset (MERFISH)                                  |
| **NSCLC**     | Phenotype Prediction  | non-small cell lung cancer (NSCLC) spatial proteomics dataset (IMC) from [Cords et al. (2024)](<https://www.cell.com/cancer-cell/fulltext/S1535-6108(23)00449-X>)                                   |

## Installation

#### Conda (Recommended)

```bash
# clone project
git clone https://github.com/BoevaLab/spatial-augmentations.git
cd spatial-augmentations

# create conda environment and install dependencies
conda env create -f environment.yaml -n myenv

# activate conda environment
conda activate myenv
```
`torch_scatter`, `torch_sparse`, and `abc_atlas` packages have to be transfered manually. The installation approaches have not been tested outside of a Linux/HPC environment. 

#### Pip

```bash
# clone project
git clone https://github.com/BoevaLab/spatial-augmentations.git
cd spatial-augmentations

# [OPTIONAL] create conda environment
conda create -n myenv python=3.9
conda activate myenv

# install pytorch according to instructions
# https://pytorch.org/get-started/

# install requirements
pip install -r requirements.txt
```

## How to run

Train model with default configuration (here, domain identification model):

```bash
# train on CPU
python src/train_domain.py trainer=cpu

# train on GPU
python src/train_domain.py trainer=gpu
```

Train model with chosen experiment configuration from [configs/experiment/](configs/experiment/):

```bash
python src/train_domain.py experiment=experiment_name.yaml
```

You can override any parameter from command line like this:

```bash
python src/train_domain.py trainer.max_epochs=20 data.batch_size=64
```

## License

Distributed under the MIT License. See [LICENSE](./LICENSE) for more information.
